### ISSUE TITLE
(reload module) doesn't reload Hy modules

### ISSUE THREAD

**Issue by ghost** [hylang/hy#712]

**Comment by Kodiologist** [comment]
A simple way to reproduce this (on Python 3) is:

    $ echo '(print "LOADED")' > test.hy
    $ hy
    hy 0.12.1+101.g5bf9ecf using CPython(default) 3.6.1 on Linux
    => (import test)
    LOADED
    => (import test)
    => (importlib.reload test)
    <module 'test' (<hy.importer.MetaLoader object at 0x7f1b0e9badd8>)>

"LOADED" should appear twice, but only appears once.

**Comment by asimjalis** [comment]
Here is a solution in Hy for reloading a file using its path. Useful for reloading from the REPL.
```clojure
(defn slurp [path] (with [f (open path)] (.read  f)))
(defn code-wrap-with-do [s] (+ "(do\n" s "\n)\n"))
(defmacro code-load [path] `(->> ~path (slurp) (code-wrap-with-do) (read-str) (eval)))
```

**Comment by minghu6** [comment]
I find It's hard to limit the scope, inner function can limit it, but can't works with `eval`.

**we may need  `let` or let `with` limit its scope.**

**Comment by Kodiologist** [comment]
Shouldn't this issue stay open? The bug remains in Hy master.

**Comment by vodik** [comment]
The right solution to this, if we want to put a reload in, is to remove the module from `sys.modules` and then import the module again.

This is what IPython's reloader does - but the reason it doesn't work with Hy is that the reloaded *explicitly* checks if `__file__` ends with ".py". Patching that, it works

Now there are a lot of issues with how Hy manipulates `sys.modules`, and they're addressed with #1518. I really need to get that done, but the patch works. Its a matter of unit test coverage and documentation that's holding it back.

Soon! I promise. My free time is at a premium at the moment for personal reasons, but I want to make it a priority.

- The importer isn't manipulating sys.modules correctly.
